libmamba/include/mamba/core/problems_graph_util.hpp

```diff
@@ -0,0 +1,143 @@
+#include <regex>
+#include <variant>
+#include <unordered_set>
+#include <sstream>
+
+#include "mamba/core/graph_util.hpp"
+#include "mamba/core/package_info.hpp"
+
+extern "C"
+{
+#include "solv/solver.h"
+}
+
+namespace mamba
+{
+
+    enum class ProblemType
+    {
+        CONFLICT,
+        NOT_FOUND,
+        NOT_INSTALLABLE,
+        BEST_NOT_INSTALLABLE,
+        ONLY_DIRECT_INSTALL,
+        EXCLUDED_BY_REPO_PRIORITY,
+        INFERIOR_ARCH,
+        PROVIDED_BY_SYSTEM
+    };
+
+    class DependencyInfo
+    {
+    public:
+        DependencyInfo(const std::string& dep);
+
+        const std::string& get_name() const;
+        const std::string& get_range() const;
+        std::string str() const;
+
+    private:
+        std::string name;
+        std::string range;
+    };
+
+    namespace NodeInfo
+    {
+        struct ResolvedPackage
+        {
+            PackageInfo m_package_info;
+            ResolvedPackage(PackageInfo const& package_info);
+        };
+
+        struct ProblematicPackage
+        {
+            std::string m_dep;
+            ProblematicPackage(std::string const& dep);
+        };
+        struct Root
+        {
+        };
+    }
+
+    namespace EdgeInfo
+    {
+        struct Require
+        {
+            DependencyInfo m_dep;
+            Require(DependencyInfo const& dep);
+        };
+        struct Constraint
+        {
+            DependencyInfo m_dep;
+            Constraint(DependencyInfo const& dep);
+        };
+    }
+
+    template <class... Ts>
+    struct overload : Ts...
+    {
+        using Ts::operator()...;
+    };
+    template <class... Ts>
+    overload(Ts...) -> overload<Ts...>;
+
+    class MNode
+    {
+    public:
+        using node_info
+            = std::variant<NodeInfo::ResolvedPackage, NodeInfo::ProblematicPackage, NodeInfo::Root>;
+        MNode(node_info const& node, std::optional<ProblemType> problem_type = std::nullopt);
+        void maybe_update_metadata(const MNode& other);
+
+    private:
+        node_info m_info;
+        std::optional<ProblemType> m_problem_type;
+
+        bool is_root() const;
+        std::string get_name() const;
+        std::optional<ProblemType> get_problem_type() const;
+    };
+
+    class MEdge
+    {
+    public:
+        using edge_info = std::variant<EdgeInfo::Require, EdgeInfo::Constraint>;
+
+        MEdge(edge_info const& info);
+        std::string get_info() const;
+
+    private:
+        edge_info m_info;
+    };
+
+    template <typename T, typename U>
+    class MProblemsGraph
+    {
+    public:
+        using graph_t = DiGraph<T, U>;
+        using node_id = typename DiGraph<T, U>::node_id;
+        using node_id_conflicts = std::unordered_map<node_id, std::vector<node_id>>;
+
+        graph_t& graph();
+        void add_conflicts(node_id node1, node_id node2);
+        node_id_conflicts const& get_conflicts() const;
+
+    private:
+        graph_t m_graph;
+        node_id_conflicts m_node_id_conflicts;
```

Comment:
Could we have that information in the graph so that we could remove the map? For instance as a fourth type of node, like so:
[![](https://mermaid.ink/img/pako:eNpFz80KgzAMB_BXKTnbou4wKJugdU-w3ayHYuss-EXXHob67qurYzkl_H-EZIFmkgooPI2YO_Qo-Yh85VWOU3IicY0wzlaWXRMSR5eUxCtiScWwH0lSB1xUxXf-4-BS78KS4Fiy56hc2DS2vW4sPm9HkoYEIhiUGYSW_qJlzzjYTg2KA_WtVK1wveXAx81TN0th1U1qOxmgrehfKgLh7HR_jw1Qa5z6oVIL_-BwqO0DbMFHHw)](https://mermaid.live/edit#pako:eNpFz80KgzAMB_BXKTnbou4wKJugdU-w3ayHYuss-EXXHob67qurYzkl_H-EZIFmkgooPI2YO_Qo-Yh85VWOU3IicY0wzlaWXRMSR5eUxCtiScWwH0lSB1xUxXf-4-BS78KS4Fiy56hc2DS2vW4sPm9HkoYEIhiUGYSW_qJlzzjYTg2KA_WtVK1wveXAx81TN0th1U1qOxmgrehfKgLh7HR_jw1Qa5z6oVIL_-BwqO0DbMFHHw)
Or as a third type of edge:
[![](https://mermaid.ink/img/pako:eNplz8EOgyAMBuBXIT0LEXYjm4m6PcGO4oEIThNRw-CwqO8-HC47rKf-6ZemXaCZlAYODyvnDokRhcqrHDNyImmNMM7WMrtQkiZnRtIVlbQqcYiE1hEXVfHJPxwdCy4uia6kCJMAmmlsh75xOzkm7G9CIQGjrZG9CrctuxPgOm20AB5apVvpBydAjFugflbS6Zvq3WSBt3J46gSkd9P9NTbAnfX6i669DK-aQ21vk59LSg)](https://mermaid.live/edit#pako:eNplz8EOgyAMBuBXIT0LEXYjm4m6PcGO4oEIThNRw-CwqO8-HC47rKf-6ZemXaCZlAYODyvnDokRhcqrHDNyImmNMM7WMrtQkiZnRtIVlbQqcYiE1hEXVfHJPxwdCy4uia6kCJMAmmlsh75xOzkm7G9CIQGjrZG9CrctuxPgOm20AB5apVvpBydAjFugflbS6Zvq3WSBt3J46gSkd9P9NTbAnfX6i669DK-aQ21vk59LSg)
Otherwise, I'd be in favor of merging this class with the `MProblemsGraphCreator`

Reply:
That is a great idea! I think having the conflict info as an edge would be the best, that would help with the merging algo

Reply:
I think we might have some issues here, we would need to change the underlying data structure from a DAG to a directed graph (that can have cycles).